src/common/internal/test_group.ts

```diff
@@ -472,11 +480,50 @@ class RunCaseSpecific implements RunCase {
           await this.beforeFn(sharedState);
         }
 
+        let allPreviousSubcasesFinalizedPromise: Promise<void> = Promise.resolve();
         if (this.subcases) {
           let totalCount = 0;
           let skipCount = 0;
+
+          // Maximum number of subcases in flight. If there are too many in flight,
+          // starting the next subcase will register `resolvePromiseBlockingSubcase`
+          // and wait until `subcaseFinishedCallback` is called.
+          const kMaxSubcasesInFlight = 500;
+          let subcasesInFlight = 0;
+          let resolvePromiseBlockingSubcase: (() => void) | undefined = undefined;
+          const subcaseFinishedCallback = () => {
+            subcasesInFlight -= 1;
+            // If there is any subcase waiting on a previous subcase to finish,
+            // unblock it now, and clear the resolve callback.
+            if (resolvePromiseBlockingSubcase) {
+              resolvePromiseBlockingSubcase();
+              resolvePromiseBlockingSubcase = undefined;
+            }
+          };
+
           for (const subParams of this.subcases) {
-            rec.info(new Error('subcase: ' + stringifyPublicParams(subParams)));
+            // Make a recorder that will defer all calls until `allPreviousSubcasesFinalizedPromise`
+            // resolves. Waiting on `allPreviousSubcasesFinalizedPromise` ensures that
+            // logs from all the previous subcases have been flushed before flushing new logs.
+            const subRec = new Proxy(rec, {
+              get: (target, k: keyof TestCaseRecorder) => {
+                const prop = TestCaseRecorder.prototype[k];
+                if (typeof prop === 'function') {
+                  return function (...args: Parameters<typeof prop>) {
+                    allPreviousSubcasesFinalizedPromise.then(() => {
+                      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                      const rv = (prop as any).apply(target, args);
+                      // Because this proxy executes functions in a deferred manner,
+                      // it should never be used for functions that need to return a value.
+                      assert(rv === undefined);
+                    });
+                  };
+                }
+                return prop;
+              },
+            });
```

Comment:
The proxy is cool but a bit hard to wrap my head around - would it be messy to implement the same thing directly on the TestCaseRecorder or a new class that wraps or subclasses it directly (just like the Proxy but as an actual class)? Other than having to manually list out all of the functions that are getting proxied here.

Reply:
Yea I'd probably just extend TestCaseRecorder and call it DeferredTestCaseRecorder and override the methods. I can do that if you'd like. I don't think it would work as nicely as the proxy though since I'd need to rebind `allPreviousSubcasesFinalizedPromise` each time it changes, or put it in a "box".

Reply:
Box seems like it would work pretty well, but leaving the proxy is probably fine, since it works.